treegp/gp_interp.py

```diff
@@ -0,0 +1,243 @@
+        else:
+            self.robust_fit = False
+
+        self.p0_robust_fit = p0
+        self.indice_meanify = indice_meanify
+
+        if isinstance(kernel,str):
+            self.kernel_template = eval_kernel(kernel)
+        else:
+            raise TypeError("kernel should be a string a list or a numpy.ndarray of string")
+
+        if self.optimizer not in ['anisotropic', 'two-pcf', 'log-likelihood', 'none']:
+            raise ValueError("Only anisotropic, two-pcf, log-likelihood and none are supported for optimizer. Current value: %s"%(self.optimizer))
+
+        if average_fits is not None:
+            import fitsio
+            average = fitsio.read(average_fits)
+            X0 = average['COORDS0'][0]
+            y0 = average['PARAMS0'][0]
+        else:
+            X0 = None
+            y0 = None
+        self._X0 = X0
+        self._y0 = y0
+
+    def _fit(self, kernel, X, y, y_err):
+        """Update the Kernel with data.
+
+        :param kernel: sklearn.gaussian_process kernel.
+        :param X:  Coordinates of the field.  (n_samples, 1 or 2)
+        :param y:  Values of the field.  (n_samples)
+        :param y_err: Error of y. (n_samples)
+        """
+        if self.optimizer != "none":
+            # Hyperparameters estimation using 2-point correlation
+            # function information.
+            if self.optimizer in ['two-pcf', 'anisotropic']:
+                anisotropic = self.optimizer == 'anisotropic'
+                self._optimizer = treegp.two_pcf(X, y, y_err,
+                                                 self.min_sep, self.max_sep,
+                                                 nbins=self.nbins,
+                                                 anisotropic=anisotropic,
+                                                 robust_fit=self.robust_fit,
+                                                 p0=self.p0_robust_fit)
+                kernel = self._optimizer.optimizer(kernel)
+            # Hyperparameters estimation using maximum likelihood fit.
+            if self.optimizer == 'log-likelihood':
+                self._optimizer = treegp.log_likelihood(X, y, y_err)
+                kernel = self._optimizer.optimizer(kernel)
+        return kernel
+
+    def predict(self, X, return_cov=False):
+        """ Predict responses to given coordinates.
+
+        :param X:  The coordinates at which to interpolate.  (n_samples, 1 or 2).
+        :returns:  Regressed parameters  (n_samples)
+        """
+        y_init = copy.deepcopy(self._y)
+        y_err = copy.deepcopy(self._y_err)
+
+        y_interp, y_cov = self.return_gp_predict(y_init-self._mean-self._spatial_average,
+                                                 self._X, X, self.kernel, y_err=y_err,
+                                                 return_cov=return_cov)
+        y_interp = y_interp.T
+        spatial_average = self._build_average_meanify(X)
+        y_interp += self._mean + spatial_average
```

Comment:
It's not clear to me why both `self._mean` _and_ `self._spatial_average` are being subtracted (added) here. In principle the spatial average is being subtracted to produce the mean, so, if I understand this, we have
```
self._spatial_average = self._build_average_meanify(X)
self._mean = np.mean(y - self._spatial_average) 
...
y_init-self._mean-self._spatial_average
```
Is subtracting the spatial average from `y_init` therefore required or redundant?

Reply:
I am not sure to understand the question, can you clarify?